node/src/components/small_network/bandwidth_limiter.rs

```diff
@@ -0,0 +1,429 @@
+    );
+}
+
+/// A handle for connections that are bandwidth limited.
+#[async_trait]
+pub(crate) trait BandwidthLimiterHandle: Send + Sync {
+    /// Waits until the sender is clear to send `num_bytes` additional bytes.
+    async fn request_allowance(&self, num_bytes: u32);
+}
+
+/// An unlimited bandwidth "limiter".
+///
+/// Does not restrict outgoing bandwidth in any way (`create_handle` returns immediately).
+#[derive(Debug)]
+pub(super) struct Unlimited;
+
+/// Handle for `Unlimited`.
+struct UnlimitedHandle;
+
+impl BandwidthLimiter for Unlimited {
+    fn create_handle(
+        &self,
+        _peer_id: NodeId,
+        _validator_id: Option<PublicKey>,
+    ) -> Box<dyn BandwidthLimiterHandle> {
+        Box::new(UnlimitedHandle)
+    }
+
+    fn update_validators(
+        &self,
+        _active_validators: HashSet<PublicKey>,
+        _upcoming_validators: HashSet<PublicKey>,
+    ) {
+    }
+}
+
+#[async_trait]
+impl BandwidthLimiterHandle for UnlimitedHandle {
+    async fn request_allowance(&self, _num_bytes: u32) {
+        // No limit.
+    }
+}
+
+/// A Bandwidth limiter dividing traffic into multiple classes based on their validator status.
+///
+/// Imposes a limit on non-validator traffic while not limiting active validator traffic at all.
+#[derive(Debug)]
+pub(super) struct ClassBasedLimiter {
+    /// Sender for commands to the limiter.
+    sender: mpsc::UnboundedSender<ClassBasedCommand>,
+}
+
+/// Bandwidth class for the `ClassBasedLimiter`.
+enum BandwidthClass {
+    /// Preferred traffic serving active validators.
+    ActiveValidator,
+    /// Traffic for upcoming validators that are not active yet.
+    UpcomingValidator,
+    /// Unclassified/low-priority traffic.
+    Bulk,
+}
+
+/// Command sent to the `ClassBasedLimiter`.
+enum ClassBasedCommand {
+    /// Updates the set of active/upcoming validators.
+    UpdateValidators {
+        /// The new set of validators active in the current era.
+        active_validators: HashSet<PublicKey>,
+        /// The new set of validators in future eras.
+        upcoming_validators: HashSet<PublicKey>,
+    },
+    /// Requests a certain amount of bandwidth.
+    RequestBandwidth {
+        /// Number of bytes requested.
+        num_bytes: u32,
+        /// Id of the requesting sender.
+        id: Arc<BandwidthConsumerId>,
+        /// Response channel.
+        responder: oneshot::Sender<()>,
+    },
+    /// Shuts down the worker task
+    Shutdown,
+}
+
+/// Handle for `ClassBasedLimiter`.
+#[derive(Debug)]
+struct ClassBasedHandle {
+    /// Sender for commands.
+    sender: mpsc::UnboundedSender<ClassBasedCommand>,
+    /// Consumer ID for the sender holding this handle.
+    consumer_id: Arc<BandwidthConsumerId>,
+}
+
+/// An identity for a bandwidth consumer.
+#[derive(Debug)]
+struct BandwidthConsumerId {
+    /// The node ID data is sent to/from.
+    peer_id: NodeId,
+    /// The remote node's `validator_id`.
+    validator_id: Option<PublicKey>,
```

Comment:
Right, so we _do_ have a way of correlating the validator public keys to node IDs? I thought we were not supposed to be able to do that, but maybe I have outdated information :wink:

Reply:
Yes, we recently switched stance on this position, as we have been moving towards this direction for quite some time.

Right now, this is strictly used for optimization though.